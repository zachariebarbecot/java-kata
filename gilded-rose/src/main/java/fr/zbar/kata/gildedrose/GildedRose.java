package fr.zbar.kata.gildedrose;

import java.util.Arrays;

public class GildedRose {
    public Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        Arrays.stream(items)
                .forEach(ItemUpdater::updateQuality);
    }
}