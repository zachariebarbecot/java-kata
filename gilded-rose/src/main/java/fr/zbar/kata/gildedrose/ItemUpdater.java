package fr.zbar.kata.gildedrose;

public class ItemUpdater {

    static void updateQuality(Item item) {
        if (item.name.equals("Sulfuras, Hand of Ragnaros")) {
            return;
        }

        item.sellIn = item.sellIn - 1;

        if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
            if (item.sellIn < 10) {
                increaseQuality(item);
            }
            if (item.sellIn < 5) {
                increaseQuality(item);
            }
            increaseQuality(item);
            if (item.sellIn < 0) {
                item.quality = 0;
            }
        } else if (item.name.equals("Aged Brie")) {
            increaseQuality(item);
            if (item.sellIn < 0) {
                increaseQuality(item);
            }
        } else {
            decreaseQuality(item);
            if (item.sellIn < 0) {
                decreaseQuality(item);
            }
        }
    }

    private static void decreaseQuality(Item item) {
        if (item.quality > 0) {
            item.quality = item.quality - 1;
        }
    }

    private static void increaseQuality(Item item) {
        if (item.quality < 50) {
            item.quality = item.quality + 1;
        }
    }
}
