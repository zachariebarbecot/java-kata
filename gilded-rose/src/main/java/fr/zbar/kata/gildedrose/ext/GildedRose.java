package fr.zbar.kata.gildedrose.ext;

import fr.zbar.kata.gildedrose.core.ItemUpdater;
import fr.zbar.kata.gildedrose.core.models.CoreItem;
import fr.zbar.kata.gildedrose.core.updater.quality.QualityUpdaterFactory;
import fr.zbar.kata.gildedrose.core.updater.sellin.SellInUpdater;

import java.util.List;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toUnmodifiableList;

public class GildedRose {

    private final ItemUpdater itemUpdater = new ItemUpdater(new QualityUpdaterFactory(), new SellInUpdater());
    private final ItemMapper itemMapper = new ItemMapper();

    public List<Item> updateQuality(List<Item> items) {
        List<CoreItem> updatedCoreItems = items.stream()
                .map(itemMapper::toCoreItem)
                .collect(Collectors.collectingAndThen(toUnmodifiableList(), itemUpdater::updateQuality));
        return updatedCoreItems.stream()
                .map(itemMapper::toItem)
                .toList();
    }
}
