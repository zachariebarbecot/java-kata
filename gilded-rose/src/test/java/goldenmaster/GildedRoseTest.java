package goldenmaster;

import fr.zbar.kata.gildedrose.GildedRose;
import fr.zbar.kata.gildedrose.Item;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

class GildedRoseTest {

    @Test
    void golden_master() {
        Item[] items = new Item[]{
                new Item("+5 Dexterity Vest", 10, 20),
                new Item("Aged Brie", 2, 0),
                new Item("Elixir of the Mongoose", 5, 7),
                new Item("Sulfuras, Hand of Ragnaros", 0, 80),
                new Item("Sulfuras, Hand of Ragnaros", -1, 80),
                new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
                new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
                new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),
                new Item("Conjured Mana Cake", 3, 6)
        };
        GildedRose app = new GildedRose(items);

        for (int i = 0; i < 7; i++) {
            app.updateQuality();
        }

        Approvals.verifyAll("", app.items);
    }

}