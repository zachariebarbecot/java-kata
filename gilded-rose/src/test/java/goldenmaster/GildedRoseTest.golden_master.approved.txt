[0] = +5 Dexterity Vest, 3, 13
[1] = Aged Brie, -5, 12
[2] = Elixir of the Mongoose, -2, 0
[3] = Sulfuras, Hand of Ragnaros, 0, 80
[4] = Sulfuras, Hand of Ragnaros, -1, 80
[5] = Backstage passes to a TAFKAL80ETC concert, 8, 29
[6] = Backstage passes to a TAFKAL80ETC concert, 3, 50
[7] = Backstage passes to a TAFKAL80ETC concert, -2, 0
[8] = Conjured Mana Cake, -4, 0
